json5/model.py

```diff
@@ -46,6 +46,28 @@ def __init__(self, *key_value_pairs, trailing_comma=None, leading_wsc=None, tok=
         assert leading_wsc is None or all(isinstance(item, str) or isinstance(item, Comment) for item in leading_wsc)
         super().__init__(key_value_pairs=key_value_pairs, trailing_comma=trailing_comma, leading_wsc=leading_wsc or [], tok=tok)
 
+    def find_key_value_pair(self, key):
+        for key_value_pair in self.key_value_pairs:
+            compare_key =  key_value_pair.key
+            if isinstance(compare_key, String):
+                if compare_key.characters == key:
+                    return key_value_pair
+            elif isinstance(compare_key, Identifier):
+                if compare_key.name == key:
+                    return key_value_pair
+            else:
+                raise NotImplementedError(f"node type not handled as object key: {compare_key}")
+        raise KeyError(key)
+
+    def __getitem__(self, key: Key) -> Value:
+        return self.find_key_value_pair(key).value
+
+    def __setitem__(self, key: Key, value: Value):
```

Comment:
I think it would be natural for people to do something like:

```python
model_object['foo'] = 'bar'
```
This, unfortunately, would lead to an error down the road because the `value` attribute of `KeyValuePair` must be of type `Value`.


It's probably reasonable to make a check to make sure the provided value is a valid value type.

```suggestion
    def __setitem__(self, key: Key, value: Value):
        if not isinstance(value, Value):
            raise TypeError(f"value argument must be of type Value, not {type(value)!r}")
```

It may also be helpful, perhaps, to automatically convert acceptable Python argument types (e.g. those supported by `modelize`).

```suggestion
    def __setitem__(self, key: Key, value: Value):
        if not isinstance(value, Value):
            try:
                _value = modelize(value)
                if not isinstance(_value, Value):
                    raise TypeError(f"value argument must be of type Value, not {type(value)!r}")
                else:
                    value = _value
            except (TypeError, ValueError, NotImplementedError): # not sure if these are the only errors modelize can raise
                raise TypeError(f"value argument must be of type Value, not {type(value)!r}")
               
```

The above suggestions are just ideas, I haven't checked that they're even syntactically valid, let alone tested they work as intended :-)

Reply:
I thought about this as well. Might be something for a follow up pull request :)

Reply:
So it might be possible to directly convert it on `__setitem__` as in your suggestion - or do it later in the serialization step. 

Neither "feels right" to me :sweat_smile: so i implemented neither.

Reply:
> Neither "feels right" to me

Yeah, I can identify with that 😅  

I had gone down this rabbit hole at least once before... got close to the beginnings of a seamless model that would have appropriate magic methods to make things like `model_object['foo'] == 'bar'` work, but ultimately some differences were unreconcilable, so it never made it in. My hope was to use that for the default loader, but it may actually be worth revisiting for the modelloader 🤔 


In any case, I agree it sounds like something that can be done in a later pull request.